README.md
Bamboo-agent Cookbook [![Build Status](https://travis-ci.org/Numergy/bamboo-agent-cookbook.svg)](https://travis-ci.org/Numergy/bamboo-agent-cookbook)
================
A Chef cookbook for managing Bamboo agents.

It can:
- Install multiple agents side-by-side on a node
- Ensure agents are running / started up after a reboot
- Set properties in a agent's wrapper.conf
- Manage agent capabilities

This cookbook is based on the Puppet module: https://github.com/kayakco/puppet-bamboo_agent

Attributes
----------
#### bamboo-agent::default
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['bamboo-agent']['server']['address']</tt></td>
    <td>String</td>
    <td>Bamboo server address</td>
    <td><tt>localhost</tt></td>
  </tr>
  <tr>
    <td><tt>['bamboo-agent']['server']['protocol']</tt></td>
    <td>String</td>
    <td>Bamboo server protocol</td>
    <td><tt>localhost</tt></td>
  </tr>
  <tr>
    <td><tt>['bamboo-agent']['server']['port']</tt></td>
    <td>Integer/String</td>
    <td>Bamboo server port</td>
    <td><tt>8085</tt></td>
  </tr>
  <tr>
    <td><tt>['bamboo-agent']['install_dir']</tt></td>
    <td>String</td>
    <td>whether to install bamboo agents</td>
    <td><tt>/usr/local/bamboo</tt></td>
  </tr>
  <tr>
    <td><tt>['bamboo-agent']['installer_jar']</tt></td>
    <td>String</td>
    <td>whether to download the installer</td>
    <td><tt>"#{node['bamboo-agent']['install_dir']}/bamboo-agent-installer.jar"</tt></td>
  </tr>
  <tr>
    <td><tt>['bamboo-agent']['user']['name']</tt></td>
    <td>String</td>
    <td>The username for bamboo agents</td>
    <td><tt>bamboo</tt></td>
  </tr>
  <tr>
    <td><tt>['bamboo-agent']['user']['group']</tt></td>
    <td>String</td>
    <td>The group for bamboo agents</td>
    <td><tt>bamboo</tt></td>
  </tr>
  <tr>
    <td><tt>['bamboo-agent']['user']['manage']</tt></td>
    <td>Boolean</td>
    <td>If the user home must be managed</td>
    <td><tt>true</tt></td>
  </tr>
  <tr>
    <td><tt>['bamboo-agent']['user']['shell']</tt></td>
    <td>String</td>
    <td>The user shell</td>
    <td><tt>/bin/bash</tt></td>
  </tr>
  <tr>
    <td><tt>['bamboo-agent']['capabilities']</tt></td>
    <td>Hash</td>
    <td>The default capabilities</td>
    <td><tt>{}</tt></td>
  </tr>
  <tr>
    <td><tt>['bamboo-agent']['agents']</tt></td>
    <td>Hash</td>
    <td>Agents to deploy</td>
    <td><tt>{}</tt></td>
  </tr>
</table>

Usage
-----
#### bamboo-agent::default

```json
{
  "bamboo-agent": {
    "server": {
      "address": "my-bamboo-url"
    },
    "agents": [
      {"id":"1", "capabilities": {"system.builder.command.Bash": "/bin/bash"}},
      {"id":"2"}
    ]
  },
  "run_list": [
    "recipe[bamboo-agent]"
  ]
}
```

Running tests
-------------
This cookbook use Rake to run tests suites:  

- First install dependencies:  
`bundle install`  

- Run all checkstyle tests:  
`bundle exec rake test:checkstyle`  

- Run all rspec tests:  
`bundle exec rake test:specs`  

- Run all kitchen tests:  
`bundle exec rake test:kitchen`  

- Run all tests (checkstyle and rspec):  
`bundle exec rake test`  

Available tests suites:  
- Knife
- RuboCop
- Foodcritic
- ChefSpec
- Kitchen

Contributing
------------

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
|                      |                                             |
|:---------------------|:--------------------------------------------|
| **Author**           | Pierre Rambaud <pierre.rambaud@numergy.com> |
| **Author**           | Antoine Rouyer <antoine.rouyer@numergy.com> |
|                      |                                             |
| **Copyright**        | Copyright (c) 2014 Numergy                  |

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
Bamboo-agent CHANGELOG
=================

This file is used to list changes made in each version of the bamboo-agent cookbook.

0.7.0
-----
- Use boxes store on http://opscode-vm-bento.s3.amazonaws.com/ for kitchen.yml
- Service bamboo agent must supports restart and status to prevent errors when start action is used
- Update init script to include LSBInitScripts (https://wiki.debian.org/LSBInitScripts)

0.6.0
-----
- Update dependencies

0.5.0
-----
- Missing build-essential cookbook

0.4.0
-----
- Licensing, and Chefspec matchers

0.3.0
-----
- Add resource and provider to allow the installation of a agent

0.2.0
-----
- Debian and CentOs support

0.1.0
-----
- Initial release of bamboo-agent



spec/spec_helper.rb
# -*- coding: utf-8 -*-
#
require 'rspec/expectations'
require 'chefspec'
require 'chefspec/berkshelf'
require 'chefspec/cacher'

ChefSpec::Coverage.start! { add_filter 'bamboo-agent' }

require 'chef/application'


spec/default_spec.rb
# -*- coding: utf-8 -*-
#
require_relative 'spec_helper'

describe 'debian=>bamboo-agent::default' do
  # speedup thing as you don't use mocking
  cached(:subject) do
    runner = ChefSpec::ServerRunner.new(platform: 'debian', version: '7.4')
    runner.converge('bamboo-agent::default')
  end

  # Write full examples using the `expect` syntax
  it 'does includes recipes' do
    expect(subject).to include_recipe('apt')
    expect(subject).to include_recipe('augeas::geminstall')
    expect(subject).to include_recipe('java')
    expect(subject).to include_recipe('build-essential')
    expect(subject).to include_recipe('bamboo-agent::download')
    expect(subject).to include_recipe('bamboo-agent::install')
  end
end

describe 'centos=>bamboo-agent::default' do
  # speedup thing as you don't use mocking
  cached(:subject) do
    runner = ChefSpec::ServerRunner.new(platform: 'centos', version: '6.0')
    runner.converge('bamboo-agent::default')
  end

  # Write full examples using the `expect` syntax
  it 'does includes recipes' do
    expect(subject).to include_recipe('yum')
    expect(subject).to include_recipe('augeas::geminstall')
    expect(subject).to include_recipe('java')
    expect(subject).to include_recipe('bamboo-agent::download')
    expect(subject).to include_recipe('bamboo-agent::install')
  end
end


spec/install_spec.rb
# -*- coding: utf-8 -*-
#
require_relative 'spec_helper'

describe 'bamboo-agent::install' do
  describe 'with wrong agent capabilities' do
    let(:subject) do
      ChefSpec::ServerRunner.new do |node|
        node.set['bamboo-agent']['agents'] = [
          {
            id: 'this is wrong id'
          }
        ]
      end.converge(described_recipe)
    end

    it 'should fail' do
      expect { subject }.to raise_error(Chef::Exceptions::ValidationFailed)
    end
  end

  describe 'with wrong agent capabilities' do
    let(:subject) do
      ChefSpec::ServerRunner.new do |node|
        node.set['bamboo-agent']['agents'] = [
          {
            id: '1',
            capabilities: 'string'
          }
        ]
      end.converge(described_recipe)
    end

    it 'should fail' do
      expect { subject }.to raise_error(Chef::Exceptions::ValidationFailed)
    end
  end

  describe 'with wrong agent wrapper_conf_properties' do
    let(:subject) do
      ChefSpec::ServerRunner.new do |node|
        node.set['bamboo-agent']['agents'] = [
          {
            id: '1',
            capabilities: {},
            wrapper_conf_properties: 'string'
          }
        ]
      end.converge(described_recipe)
    end

    it 'should fail' do
      expect { subject }.to raise_error(Chef::Exceptions::ValidationFailed)
    end
  end

  describe 'with parameters' do
    let(:subject) do
      ChefSpec::ServerRunner.new(step_into: ['bamboo_agent_install']) do |node|
        node.set['bamboo-agent']['agents'] = [
          {
            id: '1',
            capabilities: {},
            wrapper_conf_properties: {}
          }
        ]
      end.converge(described_recipe)
    end

    it 'should install agent' do
      expect(subject).to install_bamboo_agent('1')
        .with(id: '1',
              capabilities: {},
              wrapper_conf_properties: {})

      expect(subject).to create_directory('/usr/local/bamboo/agent1-home').with(
        owner: 'bamboo',
        group: 'bamboo',
        mode: '0755'
      )

      expect(subject).to run_execute('install-agent1').with(
        user: 'bamboo',
        group: 'bamboo',
        command: 'java -Dbamboo.home=/usr/local/bamboo/agent1-home -jar /usr/local/bamboo/bamboo-agent-installer.jar \'http://localhost:8085/agentServer\' install',
        creates: '/usr/local/bamboo/agent1-home/bin/bamboo-agent.sh'
      )

      expect(subject).to create_template('/etc/init.d/bamboo-agent1').with(
        source: 'init-script.erb',
        mode: '755',
        owner: 'root',
        group: 'root',
        variables: { user: 'bamboo', script: '/usr/local/bamboo/agent1-home/bin/bamboo-agent.sh', agent_id: '1' }
      )

      expect(subject).to start_service('bamboo-agent1')
      expect(subject).to enable_service('bamboo-agent1')

      expect(subject).to create_template('/usr/local/bamboo/agent1-home/bin/bamboo-capabilities.properties').with(
        user: 'bamboo',
        group: 'bamboo',
        mode: '0644',
        source: 'capabilities.properties.erb'
      )

      expect(subject).to create_file('/usr/local/bamboo/agent1-home/conf/wrapper.conf').with(
        user: 'bamboo',
        group: 'bamboo'
      )

      expect(subject).to run_augeas('update-wrapper-agent1').with(
        lens: 'CD_Properties.lns',
        incl: '/usr/local/bamboo/agent1-home/conf/wrapper.conf',
        changes: ['set /files/usr/local/bamboo/agent1-home/conf/wrapper.conf/wrapper.app.parameter.2 http://localhost:8085/agentServer/']
      )

      expect(subject).to create_cookbook_file('cd_properties.lns-agent1').with(
        source: 'augeas/lenses/cd_properties.aug',
        path: '/usr/share/augeas/lenses/dist/cd_properties.aug',
        owner: 'root',
        group: 'root',
        mode: '0644'
      )
    end
  end

  describe 'with capabilities and wrapper conf parameters' do
    let(:subject) do
      ChefSpec::ServerRunner.new(step_into: ['bamboo_agent_install']) do |node|
        node.set['bamboo-agent']['agents'] = [
          {
            id: '1',
            capabilities: {
              'system.builder.command.Bash' => '/bin/bash'
            },
            wrapper_conf_properties: {
              'wrapper.java.maxmemory' => 4096
            },
            private_tmp_dir: true
          }
        ]
      end.converge(described_recipe)
    end

    it 'should install agent' do
      expect(subject).to install_bamboo_agent('1')
        .with(id: '1',
              capabilities: {
                'system.builder.command.Bash' => '/bin/bash'
              },
              wrapper_conf_properties: {
                'wrapper.java.maxmemory' => 4096
              },
              private_tmp_dir: true)

      expect(subject).to create_directory('/usr/local/bamboo/agent1-home/.agent_tmp').with(
        owner: 'bamboo',
        group: 'bamboo',
        mode: '0755'
      )

      expect(subject).to install_package('tmpwatch')
      expect(subject).to create_cron('agent1-tmp-cleanup').with(
        command: '/usr/sbin/tmpwatch 10d /usr/local/bamboo/agent1-home/.agent_tmp'
      )

      expect(subject).to create_template('/usr/local/bamboo/agent1-home/bin/bamboo-capabilities.properties').with(
        user: 'bamboo',
        group: 'bamboo',
        mode: '0644',
        source: 'capabilities.properties.erb',
        variables: { capabilities: { 'system.builder.command.Bash' => '/bin/bash' } }
      )

      expect(subject).to create_file('/usr/local/bamboo/agent1-home/conf/wrapper.conf').with(
        user: 'bamboo',
        group: 'bamboo'
      )

      expect(subject).to run_augeas('update-wrapper-agent1').with(
        lens: 'CD_Properties.lns',
        incl: '/usr/local/bamboo/agent1-home/conf/wrapper.conf',
        changes: [
          'set /files/usr/local/bamboo/agent1-home/conf/wrapper.conf/set.TMP /usr/local/bamboo/agent1-home/.agent_tmp',
          'set /files/usr/local/bamboo/agent1-home/conf/wrapper.conf/wrapper.java.additional.3 -Djava.io.tmpdir=/usr/local/bamboo/agent1-home/.agent_tmp',
          'set /files/usr/local/bamboo/agent1-home/conf/wrapper.conf/wrapper.app.parameter.2 http://localhost:8085/agentServer/'
        ]
      )
    end
  end
end


recipes/default.rb
# -*- coding: utf-8 -*-
#
# Cookbook Name::bamboo-agent
# Recipe:: default
#
# Copyright 2014, Numergy
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

node.default['apt']['compile_time_update'] = true
case node['platform']
when 'debian', 'ubuntu'
  include_recipe 'apt'
when 'redhat', 'centos', 'fedora'
  include_recipe 'yum'
end

node.default['build-essential']['compile_time'] = true
include_recipe 'build-essential'

include_recipe 'augeas::geminstall'
include_recipe 'java'
include_recipe 'bamboo-agent::download'
include_recipe 'bamboo-agent::install'


recipes/download.rb
# -*- coding: utf-8 -*-
#
# Cookbook Name::bamboo-agent
# Recipe:: download
#
# Copyright 2014, Numergy
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

bamboo_config = node['bamboo-agent']
user = bamboo_config['user']

group user['group'] do
  action :create
end

ohai 'reload_passwd' do
  action :nothing
  plugin 'etc'
end

user user['name'] do
  supports manage_home: user['manage']
  home "/home/#{user['name']}"
  gid user['group']
  shell user['shell']
  notifies :reload, 'ohai[reload_passwd]', :immediately
  action :create
end

directory 'install-dir' do
  path bamboo_config['install_dir']
  owner user['name']
  group user['group']
  mode '0755'
  action :create
end

server_config = bamboo_config['server']
remote_file 'bamboo-agent-installer' do
  source "#{server_config['url']}/agentServer/agentInstaller"
  path bamboo_config['installer_jar']
  mode '0644'
  owner user['name']
  group user['group']
  action :create_if_missing
end


libraries/matchers.rb
# -*- coding: utf-8 -*-
#
# Cookbook Name:: bamboo-agent
# Library:: matchers
#
# Copyright 2014, Numergy
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

if defined?(ChefSpec)
  def install_bamboo_agent(resource_name)
    ChefSpec::Matchers::ResourceMatcher.new(:bamboo_agent_install, :run, resource_name)
  end
end


providers/install.rb
# -*- coding: utf-8 -*-
#
# Cookbook Name:: bamboo-agent
# Providers:: install
#
# Copyright 2014, Numergy
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
action :run do
  bamboo_config  = node['bamboo-agent']
  user           = bamboo_config['user']
  server_url     = bamboo_config['server']['url']
  home_directory = "#{bamboo_config['install_dir']}/agent#{new_resource.id}-home"
  script_path    = "#{home_directory}/bin/bamboo-agent.sh"
  service_name   = "bamboo-agent#{new_resource.id}"

  tmp_dir_props = {}
  expanded = {}
  augeas_props = []

  # merge global and agent custom capabilities
  capabilities   = new_resource.capabilities.merge(bamboo_config[:capabilities])

  directory home_directory do
    owner user['name']
    group user['group']
    mode '0755'
    action :create
  end

  execute "install-agent#{new_resource.id}" do
    environment 'PATH' => "/bin:/usr/bin:/usr/local/bin:#{ENV['PATH']}"
    user user['name']
    group user['group']
    command "java -Dbamboo.home=#{home_directory} -jar #{bamboo_config['installer_jar']} '#{server_url}/agentServer' install"
    creates script_path
    action :run
  end

  template "/etc/init.d/#{service_name}" do
    cookbook 'bamboo-agent'
    source 'init-script.erb'
    owner 'root'
    group 'root'
    mode '755'
    variables user: user['name'], script: "#{home_directory}/bin/bamboo-agent.sh", agent_id: new_resource.id
    action :create
    notifies :restart, "service[#{service_name}]", :delayed
  end

  service service_name do
    action [:enable, :start]
    supports restart: true, status: true
  end

  expand = lambda do |value|
    value.to_s.gsub('!ID!', new_resource.id.to_s)
  end

  capabilities.each_pair do |key, value|
    expanded[expand.call(key)] = expand.call(value)
  end

  template "#{home_directory}/bin/bamboo-capabilities.properties" do
    cookbook 'bamboo-agent'
    owner user['name']
    group user['group']
    mode '0644'
    variables capabilities: expanded
    source 'capabilities.properties.erb'
    action :create
  end

  if new_resource.private_tmp_dir == true
    agent_tmp = "#{home_directory}/.agent_tmp"
    directory agent_tmp do
      owner user['name']
      group user['group']
      mode '0755'
      action :create
    end

    tmp_dir_props['set.TMP'] = agent_tmp
    tmp_dir_props['wrapper.java.additional.3'] = "-Djava.io.tmpdir=#{agent_tmp}"

    package 'tmpwatch' do
      action :install
    end

    cron "agent#{new_resource.id}-tmp-cleanup" do
      command "/usr/sbin/tmpwatch 10d #{agent_tmp}"
    end
  end

  tmp_dir_props['wrapper.app.parameter.2'] = "#{server_url}/agentServer/"
  tmp_dir_props.merge(new_resource.wrapper_conf_properties)
  wrapper_path = "#{home_directory}/conf/wrapper.conf"
  tmp_dir_props.each do |key, value|
    augeas_props << "set /files#{wrapper_path}/#{key} #{value}"
  end

  cookbook_file "cd_properties.lns-agent#{new_resource.id}" do
    cookbook 'bamboo-agent'
    source 'augeas/lenses/cd_properties.aug'
    path '/usr/share/augeas/lenses/dist/cd_properties.aug'
    owner 'root'
    group 'root'
    mode '0644'
  end

  file wrapper_path do
    owner user['name']
    group user['group']
    action :create
  end

  augeas "update-wrapper-agent#{new_resource.id}" do
    lens 'CD_Properties.lns'
    incl wrapper_path
    changes augeas_props
    notifies :restart, "service[#{service_name}]", :delayed
  end

  new_resource.updated_by_last_action(true)
end
